### PR TITLE
[GPU] Fuse broadcast Multiply with last-axis ReduceSum & ReduceMean

### PR DESCRIPTION
**Pull request by dhandhalyabhavik**
## FPS Gain:

**24.94%**

### Details:

- Fuse eligible `Multiply -> ReduceSum` graphs into a two-input weighted Reduce primitive so the intermediate broadcast Multiply tensor is not materialized.
- Restrict graph fusion to static rank-4 FP16/FP32 inputs shaped as values `[B, F, Y, 16]` and weights `[B, 1, Y, 16]`, with matching batch/Y/X dimensions, `F > 1`, `Y > 1024`, reduction over the final axis, and a single Multiply consumer.
- Add `weighted_reduce_x16`, an optimized plain-`bfyx` kernel for the eligible X=16 reduction. Product materialization and accumulation order preserve the ordinary Multiply-plus-Reduce result for FP16 and FP32.
- Extend `reduce_ref` as the generic weighted fallback for blocked or padded layouts, offsets, cache serialization, and supported fused activation/quantize/eltwise post-ops.
- Keep oneDNN and existing Reduce kernels unchanged for ordinary reductions; unsupported weighted configurations are routed to the weighted reference implementation or remain as the original unfused graph.
- Add direct kernel, graph-selection, datatype, cache, fused-op, layout, padding, operand-order, dynamic-shape, rank, axis, dimension-mismatch, and multi-consumer tests.

This PR is independently buildable and does not depend on the vload8 feature-broadcast PR. Its ordinary Multiply reference tests explicitly use `generic_eltwise_ref`.

### Validation:

- Release Intel GPU plugin and `ov_gpu_unit_tests` build: passed.
- Focused weighted-reduce tests: **30/30 passed**.
- Reduce/fusion regression suite: **254/254 passed**.
- Graph-only stock baseline: **8 expected optimized-selection failures and 13 ordinary-fallback passes**. Stock retained an executed Multiply and selected no weighted kernel for positive candidates.
- Dynamic GroundingDINO model checks at B1/L6, B1/L8, and B2/L6: logits and boxes are byte-identical to stock.
- Kernel profiling: all six eligible encoder reductions select `weighted_reduce_x16`; the corresponding standalone Multiply nodes are not executed.
- Sample-image validation: raw outputs, six decoded boxes, and the rendered image are byte-identical to stock.
- A matched local end-to-end benchmark improved median throughput by **22.1%** relative to stock.
- `git diff --check`, copyright, clang-format-18, changed-line clang-tidy-18, and NCC: passed.
- Validated on two Intel devices: B60 dGPU and Panther Lake (PTL) iGPU.

### Tickets:

CVS-194431

### AI Assistance:

- **AI assistance used: yes.**
- Human validation performed: contributors defined and reviewed the required fusion, selector, and fallback behavior; the change was rebuilt on current upstream `master`; focused and regression tests, stock-versus-patched graph checks, byte-exact model and sample comparisons, kernel profiling, formatting/static-analysis checks, and repeated relative performance measurements were run locally. The contributors take responsibility for the submitted change.

### CODE REVIEW

**Comment by p-durandin** [comment]
build_jenkins

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/reduce_ref.cl [hunk 1]
```diff
@@ -183,53 +185,65 @@ KERNEL(reduce_ref)(
                                     const uint input_idx = INPUT0_GET_INDEX(bi, fi, yi, xi);
 
 #endif
+#if WEIGHTED_REDUCE
+                                    const uint weight_idx = INPUT1_GET_INDEX_SAFE(bi, fi, yi, xi);
+#    if INPUT0_TYPE_SIZE == 2
+                                    volatile ushort product_bits = as_ushort(data[input_idx] * weights[weight_idx]);
+                                    const INPUT0_TYPE reduce_value = as_half(product_bits);
```
Comment (v-Golubev) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/reduce_ref.cl]: Minor: this round-trips the product through its raw bits via a `volatile` temporary (same pattern in `weighted_reduce_x16.cl`). Presumably this forces truncation to the target precision so the fused kernel bit-matches the unfused Multiply+Reduce reference, but nothing says so. A one-line comment on why the `volatile`+`as_*` round-trip is needed (rather than just `data[input_idx] * weights[weight_idx]`) would save the next person from "simplifying" it away.

File: src/plugins/intel_gpu/src/plugin/ops/reduce.cpp [hunk 2]
```diff
@@ -43,13 +42,40 @@ static void CreateReduceOp(ProgramBuilder& p, const std::shared_ptr<ov::Node>& o
...
+    if (mode == cldnn::reduce_mode::sum && axes.size() == 1 && axes[0] == rank - 1) {
+        auto src = op->get_input_node_shared_ptr(0);
+        auto multiply = ov::as_type_ptr<ov::op::v1::Multiply>(src);
+        if (multiply && multiply->output(0).get_target_inputs().size() == 1 && multiply->get_input_partial_shape(0).is_static() &&
+            multiply->get_input_partial_shape(1).is_static()) {
+            const auto input_type = multiply->get_input_element_type(0);
+            const bool supported_type = input_type == multiply->get_input_element_type(1) && (input_type == ov::element::f16 || input_type == ov::element::f32);
+            const auto shape0 = multiply->get_input_shape(0);
+            const auto shape1 = multiply->get_input_shape(1);
+            if (supported_type && shape0.size() == 4 && shape1.size() == 4 && shape0[0] == shape1[0] && shape0[2] == shape1[2] && shape0[2] > 1024 &&
```
Comment (v-Golubev) [inline src/plugins/intel_gpu/src/plugin/ops/reduce.cpp]: This eligibility predicate (matching dtypes, rank 4, axis 3, `X==16`, `Y>1024`, one side `F==1`/other `F>1`, matching batch) is independently reimplemented in `IsWeightedReduceCandidate()` in `eltwise.cpp`, and again (in a different shape) in `ReduceKernelRef::Validate` and `ReduceKernelWeightedX16::Validate` — 4 copies total.

This is riskier than the usual copy-paste case: `CreateMultiplyOp` uses its copy to decide whether to *skip creating* the `Multiply` primitive, while this function uses its own copy to decide whether to *consume* the fused inputs. If the two ever drift, `CreateReduceOp` would fall back to `p.GetInputInfo(op)` for a `Multiply` primitive that was never added to the topology — a graph-construction failure, not just a maintenance smell.

Let's factor the shared predicate into one function (taking the `Multiply` and `ReduceSum` nodes) that both `CreateMultiplyOp` and `CreateReduceOp` call, so the two decisions can never disagree. Could you please double check whether the kernel-selector-level duplication (ref vs. weighted_x16 `Validate`) can be reduced similarly, or whether it's guaranteed redundant given the pre-fusion shape guarantees?

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/reduce_ref.cl (hunk 1 above)
Comment (v-Golubev) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/reduce_ref.cl]: Minor: this round-trips the product through its raw bits via a `volatile` temporary (same pattern in `weighted_reduce_x16.cl`). Presumably this forces truncation to the target precision so the fused kernel bit-matches the unfused Multiply+Reduce reference, but nothing says so. A one-line comment on why the `volatile`+`as_*` round-trip is needed (rather than just `data[input_idx] * weights[weight_idx]`) would save the next person from "simplifying" it away.

File: src/plugins/intel_gpu/src/graph/impls/ocl/reduce.cpp [hunk 3]
```diff
@@ -83,10 +83,13 @@ struct reduce_impl : typed_primitive_impl_ocl<reduce> {
     static kernel_params_t get_kernel_params(const kernel_impl_params& impl_param, bool is_shape_agnostic = false) {
         const auto& primitive = impl_param.typed_desc<reduce>();
         auto params = get_default_params<kernel_selector::reduce_params>(impl_param, is_shape_agnostic);
+        if (primitive->weighted)
+            params.inputs.push_back(convert_data_tensor(impl_param.get_input_layout(1)));
```
Comment (v-Golubev) [inline src/plugins/intel_gpu/src/graph/impls/ocl/reduce.cpp]: Shouldn't we firstly assert that 1st input exists?

**Review by v-Golubev — COMMENTED** [review]
The optimization idea is clear, but the implementation looks too narrow and restrictive: it works only with static shapes, there are some dimension values' requirements, only ReduceSum is supported, etc, And these checks have to be duplicated across 4 places.

Usually, such problems are solved by fusing eltwise as a post-op on its producer via the existing fused_ops machinery, rather than fusing the eltwise as pre-op of some operation. Was that considered? Could you also share the actual Multiply -> ReduceSum subgraph (with producers of values/weights) so we can see whether post-op fusion would apply here?

Also, there is two independent effects which are bundled into one performance benefit number: (1) fusing the Multiply as pre-op and (2) replacing whatever previously executed the reduction with the vectorized weighted_reduce_x16 kernel. Could you break down how much of the ~25% perf boost each contributes?

**Comment by dhandhalyabhavik** [comment]
Hi @v-Golubev, thank you for the detailed comments. 

I have divided them into the following questions.

### Q1: Why narrow, restrictive, and limited to static shapes?

This optimization targets the specific high-cost pattern found in GroundingDINO. After shape specialization (image 800x1333), these nodes have static rank-4 shapes with `X=16`, allowing safe vectorized indexing and bit-identical results.

Dynamic or unsupported shapes continue using the existing generic path. A fully generic `Multiply + ReduceSum` fusion would require broader support for axes, layouts, broadcasts, and datatypes, so I suggest handling that separately.

I agree that the repeated checks should be consolidated. The graph-level checks can use one shared helper, while kernel-selector validation should remain as an independent safety check.

### Q2: Why not use the existing `fused_ops` mechanism?

We considered it. The immediate Multiply producers are `Reshape`, backed by `Concat` and `Transpose`, which are not supported by the existing eltwise post-op fusion.

Also, producer-side `fused_ops` would still write the full multiplied tensor for the following ReduceSum. The weighted-reduce path performs multiplication and reduction together, avoiding that large intermediate tensor completely. (6x 185MB write saved for GroundingDINO work)

### Q3: What is the actual subgraph?

At a high level:

image

```text
Concat → Reshape [8,32,22223,16]
                             × → ReduceSum(axis=-1)
Transpose → Reshape [8,1,22223,16]
```

This pattern appears once in each of the **six** GroundingDINO encoder self-attention layers.

### Q4: How much does each optimization contribute?

I ran a matched four-way benchmark on PTL-86 using four balanced 30-second runs:

| Variant | Change |
|---|---:|
| Separate generic Multiply + ReduceSum | Baseline |
| Separate vload8 Multiply + ReduceSum | +3.74% |
| Fused Multiply + ReduceSum using `reduce_ref` | +19.88% |
| Fused using `weighted_reduce_x16` | +7.52% over fused `reduce_ref`, +28.90% overall |

Using the absolute throughput increase:

- The fused-reference path contributes approximately **68.8%** of the total gain.
- The specialized x16 kernel contributes the remaining **31.2%**.

The improvement is significant because this large Multiply-and-reduction pattern is executed in all six encoder layers. All four benchmark variants produced byte-identical outputs.

**Comment by deepaks2** [comment]
Fix the Unit test failure

File: src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp [hunk 4]
```diff
@@ -103,7 +105,41 @@ static void CreateAddOp(ProgramBuilder& p, const std::shared_ptr<ov::op::v1::Add
     CreateElementwiseOp(p, op, cldnn::eltwise_mode::sum);
 }
 
+static bool IsWeightedReduceCandidate(const std::shared_ptr<ov::op::v1::Multiply>& multiply) {
```
Comment (deepaks2) [inline src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp]: I think the Fusion eligibility is implemented twice with divergent-consequence risk. IsWeightedReduceCandidate()  in  plugin/ops/eltwise.cpp and the inline check in  CreateReduceOp  in  plugin/ops/reduce.cpp, I think both are doing same decision.  Just check once

File: src/plugins/intel_gpu/src/kernel_selector/kernels/reduce/reduce_kernel_weighted_x16.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        DO_NOT_USE_THIS_KERNEL(p.layerID);
+    }
+
+    if (values.X().v != 16 || weights.X().v != 16 || output.X().v != 1 || values.Y().v <= 1024 || values.Batch().v != weights.Batch().v ||
+        values.Batch().v != output.Batch().v || weights.Feature().v != 1 || values.Feature().v != output.Feature().v || values.Y().v != weights.Y().v ||
+        values.Y().v != output.Y().v) {
+        DO_NOT_USE_THIS_KERNEL(p.layerID);
+    }
+
+    return true;
```
Comment (deepaks2) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/reduce/reduce_kernel_weighted_x16.cpp:43]: I see missing mixed-dtype guard in the ReduceKernelWeightedX16::Validate.  ReduceKernelRef::Validate  rejects mixed input dtypes, but the x16 kernel accepts  F16×F32  combinations but never compares them.  Add  values.GetDType() != weights.GetDType()  to the reject condition

**Comment by clee30** [comment]
build_jenkins

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/weighted_reduce_x16.cl [hunk 6]
```diff
@@ -0,0 +1,44 @@
+// Copyright (C) 2018-2026 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "include/batch_headers/fetch_data.cl"
+
+KERNEL(weighted_reduce_x16)(const __global INPUT0_TYPE* values, const __global INPUT1_TYPE* weights, __global OUTPUT_TYPE* output) {
+    const uint gid = (uint)get_global_id(0);
+    const uint y = gid % OUTPUT_SIZE_Y;
```
Comment (clee30) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/weighted_reduce_x16.cl:12]: May be add a guard here.
if (gid >= OUTPUT_LENGTH)
    return;

File: src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp (hunk 4 above)
Comment (dhandhalyabhavik) [reply src/plugins/intel_gpu/src/plugin/ops/eltwise.cpp]: outdated

already fixed by shared `weighted_reduce.hpp` helper. No further work.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/reduce/reduce_kernel_weighted_x16.cpp (hunk 5 above)
Comment (dhandhalyabhavik) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/reduce/reduce_kernel_weighted_x16.cpp:43]: added a fix.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/reduce/reduce_kernel_ref.cpp [hunk 7]
```diff
@@ -28,9 +28,39 @@ ParamsKey ReduceKernelRef::GetSupportedKey() const {
...
+    const auto& weights = params.inputs[1];
+    const auto& output = params.outputs[0];
+    if (values.Dimentions() != 4 || weights.Dimentions() != 4 || output.Dimentions() != 4 ||
+        (values.GetDType() != Datatype::F16 && values.GetDType() != Datatype::F32) || values.GetDType() != weights.GetDType() || values.X().v != 16 ||
+        weights.X().v != 16 || output.X().v != 1 || values.Y().v <= 1024 || values.Batch().v != weights.Batch().v || values.Batch().v != output.Batch().v ||
+        weights.Feature().v != 1 || values.Feature().v <= 1 || values.Feature().v != output.Feature().v || values.Y().v != weights.Y().v ||
+        values.Y().v != output.Y().v) {
+        DO_NOT_USE_THIS_KERNEL(p.layerID);
+    }
+
```
Comment (clee30) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/reduce/reduce_kernel_ref.cpp:39]: There are similar checks in the [reduce_kernel_weighted_x16.cpp](https://github.com/openvinotoolkit/openvino/pull/38038/changes#diff-af5501560186002f870c955b83f6a9c6c770ad7e17fca3acb6a5acd603e791d0)

Probably add like IsWeightedReducePattern(const reduce_params& params) in reduce_kernel_base.cpp to reduce the duplicate code

Comment (dhandhalyabhavik) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/reduce/reduce_kernel_ref.cpp:39]: removed duplicity.

File: src/plugins/intel_gpu/include/intel_gpu/plugin/weighted_reduce.hpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+struct WeightedReduceMatch {
+    std::shared_ptr<ov::op::v1::Multiply> multiply;
+    size_t values_idx = 0;
+    size_t weights_idx = 1;
+};
+
+/// @brief Returns the operands of an eligible Multiply-ReduceSum weighted reduction.
+inline std::optional<WeightedReduceMatch> get_weighted_reduce_match(const ov::Node* reduce_node) {
+    if (reduce_node == nullptr || !ov::is_type<ov::op::v1::ReduceSum>(reduce_node)) {
+        return std::nullopt;
```
Comment (clee30) [inline src/plugins/intel_gpu/include/intel_gpu/plugin/weighted_reduce.hpp:31]: It is good to extend to support mean also for some pose, detection, keypoint models that compute softmax-weighted expectation over bins.   The code to add mean support should be easy for existing kernel.

**Review by deepaks2 — APPROVED** [review]

File: src/plugins/intel_gpu/include/intel_gpu/plugin/weighted_reduce.hpp (hunk 8 above)
Comment (dhandhalyabhavik) [reply src/plugins/intel_gpu/include/intel_gpu/plugin/weighted_reduce.hpp:31]: Added, please review.

**Comment by clee30** [comment]
build_jenkins
